kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoWindowedTable.java

```diff
@@ -0,0 +1,472 @@
+import com.mongodb.client.model.UpdateOptions;
+import com.mongodb.client.model.Updates;
+import com.mongodb.client.model.WriteModel;
+import com.mongodb.client.result.UpdateResult;
+import dev.responsive.kafka.internal.db.MongoWindowFlushManager;
+import dev.responsive.kafka.internal.db.RemoteWindowedTable;
+import dev.responsive.kafka.internal.db.partitioning.SegmentPartitioner;
+import dev.responsive.kafka.internal.db.partitioning.SegmentPartitioner.SegmentPartition;
+import dev.responsive.kafka.internal.stores.RemoteWriteResult;
+import dev.responsive.kafka.internal.utils.Iterators;
+import dev.responsive.kafka.internal.utils.WindowedKey;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.errors.TaskMigratedException;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.bson.codecs.configuration.CodecProvider;
+import org.bson.codecs.configuration.CodecRegistry;
+import org.bson.codecs.pojo.PojoCodecProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MongoWindowedTable implements RemoteWindowedTable<WriteModel<WindowDoc>> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MongoWindowedTable.class);
+  private static final String METADATA_COLLECTION_NAME = "window_metadata";
+
+  private static final UpdateOptions UPSERT_OPTIONS = new UpdateOptions().upsert(true);
+
+  private final String name;
+  private final SegmentPartitioner partitioner;
+
+  // whether to put windowStartMs first in the composite windowed key format in WindowDoc
+  private final boolean timestampFirstOrder;
+
+  private final MongoDatabase database;
+  private final MongoCollection<WindowMetadataDoc> metadata;
+  private final ConcurrentMap<Integer, PartitionSegments> kafkaPartitionToSegments =
+      new ConcurrentHashMap<>();
+
+  private static class PartitionSegments {
+    private final MongoDatabase database;
+    private final SegmentPartitioner partitioner;
+    private final long epoch;
+
+    // Recommended to keep the total number of collections under 10,000, so we should not
+    // let num_segments * num_kafka_partitions exceed 10k at the most
+    private final Map<SegmentPartition, MongoCollection<WindowDoc>> segmentWindows;
+
+    public PartitionSegments(
+        final MongoDatabase database,
+        final SegmentPartitioner partitioner,
+        final int kafkaPartition,
+        final long streamTime,
+        final long epoch
+    ) {
+      this.database = database;
+      this.partitioner = partitioner;
+      this.epoch = epoch;
+      this.segmentWindows = new ConcurrentHashMap<>();
+
+      final var activeSegments = partitioner.activeSegments(kafkaPartition, streamTime);
+      if (activeSegments.isEmpty()) {
```

Comment:
would it make sense to do this from the base flush manager instead so all the segment creation/deletion logic is in one place?